crates/runc-rust/src/io.rs

```diff
@@ -0,0 +1,333 @@
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+use std::fmt::{self, Debug, Formatter};
+use std::fs::File;
+use std::os::unix::io::FromRawFd;
+use std::os::unix::prelude::{AsRawFd, RawFd};
+use std::process::Command;
+use std::sync::Mutex;
+
+use nix::fcntl::OFlag;
+use nix::sys::stat::Mode;
+use nix::unistd::{Gid, Uid};
+
+pub trait RuncIO: Sync + Send {
+    /// Return write side of stdin
+    fn stdin(&self) -> Option<File> {
+        None
+    }
+
+    /// Return read side of stdout
+    fn stdout(&self) -> Option<File> {
+        None
+    }
+
+    /// Return read side of stderr
+    fn stderr(&self) -> Option<File> {
+        None
+    }
+
+    fn close(&self) {
+        unimplemented!()
+    }
+
+    /// Set IO for passed command.
+    /// Read side of stdin, write side of stdout and write side of stderr should be provided to command.
+    fn set(&self, _cmd: &mut Command) -> std::io::Result<()> {
+        unimplemented!()
+    }
+
+    // tokio version of set()
+    fn set_tk(&self, _cmd: &mut tokio::process::Command) -> std::io::Result<()> {
+        unimplemented!()
+    }
+
+    fn close_after_start(&self) {
+        unimplemented!()
+    }
+}
+
+impl Debug for dyn RuncIO {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "RuncIO",)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct IOOption {
+    pub open_stdin: bool,
+    pub open_stdout: bool,
+    pub open_stderr: bool,
+}
+
+impl Default for IOOption {
+    fn default() -> Self {
+        Self {
+            open_stdin: true,
+            open_stdout: true,
+            open_stderr: true,
+        }
+    }
+}
+
+/// This struct represents pipe that can be used to transfer
+/// stdio inputs and outputs
+/// when one side of closed, this struct represent it with [`None`]
+#[derive(Debug)]
+pub struct Pipe {
+    // Might be ugly hack: using mutex in order to take rd/wr under immutable [`Pipe`]
+    rd: Mutex<Option<File>>,
+    wr: Mutex<Option<File>>,
+}
+
+impl Pipe {
+    pub fn new() -> std::io::Result<Self> {
+        let (r, w) = nix::unistd::pipe()?;
+        let (rd, wr) = unsafe {
+            (
+                Mutex::new(Some(File::from_raw_fd(r))),
+                Mutex::new(Some(File::from_raw_fd(w))),
+            )
+        };
+        Ok(Self { rd, wr })
+    }
+
+    pub fn take_read(&self) -> Option<File> {
+        let mut m = self.rd.lock().unwrap();
+        m.take()
+    }
+
+    pub fn take_write(&self) -> Option<File> {
+        let mut m = self.wr.lock().unwrap();
+        m.take()
+    }
+
+    pub fn close_read(&self) {
+        let mut m = self.rd.lock().unwrap();
+        let _ = m.take();
+    }
+
+    pub fn close_write(&self) {
+        let mut m = self.wr.lock().unwrap();
+        let _ = m.take();
+    }
```

Comment:
lock() returns Result. Does it make sense to return the errors rather than panic?

Reply:
not in case of std `Mutex` - https://users.rust-lang.org/t/should-i-unwrap-a-mutex-lock/61519